File: backend/app/modules/fleet/api/fuel_pumps.py
"""Fuel pump endpoints."""

from decimal import Decimal

from fastapi import APIRouter

from app.shared.api.deps import CurrentUserDep, DBDep
from app.models.enums import FuelType
from app.schemas import (
    FuelPumpCreate,
    FuelPumpLevelAdjustment,
    FuelPumpResponse,
    FuelPumpUpdate,
)
from app.services import fuel_pump_service

router = APIRouter()


@router.get("", response_model=list[FuelPumpResponse])
def list_fuel_pumps(
    db: DBDep,
    current_user: CurrentUserDep,
    skip: int = 0,
    limit: int = 100,
) -> list[FuelPumpResponse]:
    """List all fuel pumps for the organization."""
    pumps = fuel_pump_service.get_multi(
        db,
        skip=skip,
        limit=limit,
        organization_id=current_user.organization_id,
    )
    return [FuelPumpResponse.model_validate(p) for p in pumps]


@router.post("", response_model=FuelPumpResponse, status_code=201)
def create_fuel_pump(
    pump_in: FuelPumpCreate,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpResponse:
    """Create a new fuel pump."""
    pump_data = pump_in.model_copy(
        update={"organization_id": current_user.organization_id}
    )
    pump = fuel_pump_service.create(db, obj_in=pump_data)
    return FuelPumpResponse.model_validate(pump)


@router.get("/alerts", response_model=list[FuelPumpResponse])
def list_pumps_with_alerts(
    db: DBDep,
    current_user: CurrentUserDep,
) -> list[FuelPumpResponse]:
    """List fuel pumps with active alerts (low level or maintenance due)."""
    pumps = fuel_pump_service.get_with_alerts(
        db,
        organization_id=current_user.organization_id,
    )
    return [FuelPumpResponse.model_validate(p) for p in pumps]


@router.get("/stats")
def get_pump_stats(
    db: DBDep,
    current_user: CurrentUserDep,
) -> dict:
    """Get fuel pump statistics for the organization."""
    return fuel_pump_service.get_stats(
        db,
        organization_id=current_user.organization_id,
    )


@router.get("/by-fuel-type/{fuel_type}", response_model=list[FuelPumpResponse])
def list_pumps_by_fuel_type(
    fuel_type: FuelType,
    db: DBDep,
    current_user: CurrentUserDep,
    active_only: bool = True,
) -> list[FuelPumpResponse]:
    """List fuel pumps by fuel type (useful for POS pump selection)."""
    pumps = fuel_pump_service.get_by_fuel_type(
        db,
        organization_id=current_user.organization_id,
        fuel_type=fuel_type,
        active_only=active_only,
    )
    return [FuelPumpResponse.model_validate(p) for p in pumps]


@router.get("/{pump_id}", response_model=FuelPumpResponse)
def get_fuel_pump(
    pump_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpResponse:
    """Get a fuel pump by ID."""
    pump = fuel_pump_service.get_or_404(db, pump_id)
    return FuelPumpResponse.model_validate(pump)


@router.patch("/{pump_id}", response_model=FuelPumpResponse)
def update_fuel_pump(
    pump_id: str,
    pump_in: FuelPumpUpdate,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpResponse:
    """Update a fuel pump."""
    pump = fuel_pump_service.get_or_404(db, pump_id)
    updated = fuel_pump_service.update(db, db_obj=pump, obj_in=pump_in)
    return FuelPumpResponse.model_validate(updated)


@router.delete("/{pump_id}", status_code=204)
def delete_fuel_pump(
    pump_id: str,
    db: DBDep,
    current_user: CurrentUserDep,
) -> None:
    """Delete a fuel pump (soft delete)."""
    fuel_pump_service.delete(db, pump_id)


@router.post("/{pump_id}/adjust-level", response_model=FuelPumpResponse)
def adjust_pump_level(
    pump_id: str,
    adjustment: FuelPumpLevelAdjustment,
    db: DBDep,
    current_user: CurrentUserDep,
) -> FuelPumpResponse:
    """Manually adjust pump level (positive to add, negative to remove)."""
    pump = fuel_pump_service.adjust_level(
        db,
        pump_id=pump_id,
        adjustment=adjustment.adjustment,
    )
    return FuelPumpResponse.model_validate(pump)

File: frontend/src/lib/api/fuel-pumps.ts
/**
 * Fuel Pumps API Service
 */

import type { FuelPump, FuelPumpStats, FuelType } from "@/types";
import { apiClient } from "./client";
import type {
  FuelPumpApiResponse,
  FuelPumpCreateRequest,
  FuelPumpUpdateRequest,
  FuelPumpLevelAdjustmentRequest,
  FuelPumpStatsResponse,
  ListParams,
} from "./types";
import { transformFuelPump, transformFuelPumps, transformFuelPumpStats } from "./transformers";

export interface FuelPumpListParams extends ListParams {
  fuel_type?: string;
  status?: string;
}

export const fuelPumpsService = {
  /**
   * Get all fuel pumps
   */
  async getAll(params?: FuelPumpListParams): Promise<FuelPump[]> {
    const queryParams: Record<string, string | number | boolean | undefined> = {
      skip: params?.skip,
      limit: params?.limit,
      fuel_type: params?.fuel_type,
      status: params?.status,
    };

    const response = await apiClient.get<FuelPumpApiResponse[]>("/fuel-pumps", {
      params: queryParams,
    });
    return transformFuelPumps(response);
  },

  /**
   * Get fuel pump by ID
   */
  async getById(id: string): Promise<FuelPump> {
    const response = await apiClient.get<FuelPumpApiResponse>(`/fuel-pumps/${id}`);
    return transformFuelPump(response);
  },

  /**
   * Get fuel pumps by fuel type (useful for POS pump selection)
   */
  async getByFuelType(fuelType: FuelType, activeOnly: boolean = true): Promise<FuelPump[]> {
    const response = await apiClient.get<FuelPumpApiResponse[]>(
      `/fuel-pumps/by-fuel-type/${fuelType}`,
      { params: { active_only: activeOnly } }
    );
    return transformFuelPumps(response);
  },

  /**
   * Get fuel pumps with active alerts (low level or maintenance due)
   */
  async getAlerts(): Promise<FuelPump[]> {
    const response = await apiClient.get<FuelPumpApiResponse[]>("/fuel-pumps/alerts");
    return transformFuelPumps(response);
  },

  /**
   * Get pump statistics
   */
  async getStats(): Promise<FuelPumpStats> {
    const response = await apiClient.get<FuelPumpStatsResponse>("/fuel-pumps/stats");
    return transformFuelPumpStats(response);
  },

  /**
   * Create new fuel pump
   */
  async create(data: FuelPumpCreateRequest): Promise<FuelPump> {
    const response = await apiClient.post<FuelPumpApiResponse>("/fuel-pumps", data);
    return transformFuelPump(response);
  },

  /**
   * Update fuel pump
   */
  async update(id: string, data: FuelPumpUpdateRequest): Promise<FuelPump> {
    const response = await apiClient.patch<FuelPumpApiResponse>(`/fuel-pumps/${id}`, data);
    return transformFuelPump(response);
  },

  /**
   * Delete fuel pump (soft delete)
   */
  async delete(id: string): Promise<void> {
    await apiClient.delete(`/fuel-pumps/${id}`);
  },

  /**
   * Manually adjust pump level
   */
  async adjustLevel(id: string, adjustment: number): Promise<FuelPump> {
    const request: FuelPumpLevelAdjustmentRequest = { adjustment };
    const response = await apiClient.post<FuelPumpApiResponse>(
      `/fuel-pumps/${id}/adjust-level`,
      request
    );
    return transformFuelPump(response);
  },
};
